### ISSUE TITLE
Git filter interrogation issue

### ISSUE THREAD

**Issue by vidartf** [jupyter/nbdime#415]
Originally reported in https://github.com/kynan/nbstripout/issues/44#issuecomment-412489658, there seems to be an issue getting the correct filter name from the config.

<details><summary>Reported trace</summary>
<pre>
Traceback (most recent call last):
File "/Users/trtim/miniconda3/envs/nbdime/bin/nbdiff", line 11, in sys.exit(main())
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/site-packages/nbdime/nbdiffapp.py", line 128, in main return main_diff(arguments)
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/site-packages/nbdime/nbdiffapp.py", line 39, in main_diff for fbase, fremote in changed_notebooks(base, remote, paths):
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/site-packages/nbdime/gitfiles.py", line 160, in changed_notebooks entry.b_path, entry.b_blob, ref_remote, repo_dir)
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/site-packages/nbdime/gitfiles.py", line 110, in _get_diff_entry_stream ret = apply_possible_filter(path)
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/site-packages/nbdime/vcs/git/filter_integration.py", line 80, in apply_possible_filter stderr=STDOUT, shell=USE_SHELL
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/subprocess.py", line 336, in check_output **kwargs).stdout
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/subprocess.py", line 403, in run with Popen(*popenargs, **kwargs) as process:
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/subprocess.py", line 709, in __init__ restore_signals, start_new_session)
File "/Users/trtim/miniconda3/envs/nbdime/lib/python3.6/subprocess.py", line 1344, in _execute_child raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '"/Users/trtim/miniconda3/envs/nbdime/bin/python3.6"': '"/Users/trtim/miniconda3/envs/nbdime/bin/python3.6"' 
</pre>
</details>
</br>

@timtroendle What is the output of running `git check-attr -z filter <notebook path here>` for the notebook in question?

**Comment by timtroendle** [comment]
`git check-attr -z filter Untitled.ipynb` -> `Untitled.ipynbfilternbstripout`

**Comment by vidartf** [comment]
And then the output of `git config --get -z filter.nbstripout.clean` ?

**Comment by timtroendle** [comment]
`~/miniconda3/envs/nbdime/bin/nbstripout`

**Comment by vidartf** [comment]
Would you mind running the following python code?

```python
from subprocess import check_output, STDOUT

path = "Untitled.ipynb"
filter_cmd = "~/miniconda3/envs/nbdime/bin/nbstripout"
with io.open(path, 'r', encoding="utf8") as f:
    output = check_output(
        filter_cmd.split(),
        stdin=f,
        stderr=STDOUT, shell=False
    ).decode('utf8', 'replace')
print (output)
```

And then, run it once with `shell=True` for reference.

**Comment by timtroendle** [comment]
shell=False

<details><summary>Full trace</summary>
<pre>
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-4-07f2044cec00> in <module>()
      7         filter_cmd.split(),
      8         stdin=f,
----> 9         stderr=STDOUT, shell=False
     10     ).decode('utf8', 'replace')
     11 print (output)

~/miniconda3/envs/nbdime/lib/python3.6/subprocess.py in check_output(timeout, *popenargs, **kwargs)
    334 
    335     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
--> 336                **kwargs).stdout
    337 
    338 

~/miniconda3/envs/nbdime/lib/python3.6/subprocess.py in run(input, timeout, check, *popenargs, **kwargs)
    401         kwargs['stdin'] = PIPE
    402 
--> 403     with Popen(*popenargs, **kwargs) as process:
    404         try:
    405             stdout, stderr = process.communicate(input, timeout=timeout)

~/miniconda3/envs/nbdime/lib/python3.6/subprocess.py in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, encoding, errors)
    707                                 c2pread, c2pwrite,
    708                                 errread, errwrite,
--> 709                                 restore_signals, start_new_session)
    710         except:
    711             # Cleanup if the child failed starting.

~/miniconda3/envs/nbdime/lib/python3.6/subprocess.py in _execute_child(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, start_new_session)
   1342                         if errno_num == errno.ENOENT:
   1343                             err_msg += ': ' + repr(err_filename)
-> 1344                     raise child_exception_type(errno_num, err_msg, err_filename)
   1345                 raise child_exception_type(err_msg)
   1346 

FileNotFoundError: [Errno 2] No such file or directory: '~/miniconda3/envs/nbdime/bin/nbstripout': '~/miniconda3/envs/nbdime/bin/nbstripout'
</pre>
</details>

shell=True

<details><summary>Full trace</summary>
<pre>
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "4 + 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
</pre>
</details>

**Comment by vidartf** [comment]
@timtroendle What platform/OS are you on? Python version?

@minrk Any idea why [this code](https://github.com/jupyter/nbdime/blob/3d7d95e04ee641494de4c4eab0e8d7475ff1d309/nbdime/vcs/git/filter_integration.py#L80) seems to require `shell=True` for this case?

**Comment by timtroendle** [comment]
macOS 10.13.6

Python 3.6.6

**Comment by vidartf** [comment]
@kynan Thanks! It seems weird that subprocess does not handle `~`, but if that's all it takes I'll patch it.

**Comment by vidartf** [comment]
As far as I can tell from [reading the git source](https://github.com/git/git/blob/ffc6fa0e396238de3a30623912980263b4f283ab/convert.c#L668), the filter command is run with shell. Does this argue for always using `shell=True` here?

**Comment by kynan** [comment]
I think always using the shell would make sense, yes

**Comment by vidartf** [comment]
Released in 1.0.3.
